Loop/Managers/DeviceDataManager.swift

```diff
@@ -1347,119 +1348,176 @@ extension Notification.Name {
-        try await self.enactBolus(units: validBolusAmount, activationType: .manualNoRecommendation)
+    func deliverBolus(amountInUnits: Double) async throws {
+        
+        guard amountInUnits > 0 else {
+            throw BolusActionError.invalidBolus
+        }
+        
+        guard let maxBolusAmount = loopManager.settings.maximumBolus else {
+            throw BolusActionError.missingMaxBolus
+        }
+        
+        guard amountInUnits <= maxBolusAmount else {
+            throw BolusActionError.exceedsMaxBolus
+        }
+        
+        try await self.enactBolus(units: amountInUnits, activationType: .manualNoRecommendation)
         await remoteDataServicesManager.triggerUpload(for: .dose)
-        self.analyticsServicesManager.didBolus(source: "Remote", units: validBolusAmount)
+        self.analyticsServicesManager.didBolus(source: "Remote", units: amountInUnits)
     }
     
-    //Remote Carb Entry
+    enum BolusActionError: LocalizedError {
+        
+        case invalidBolus
+        case missingMaxBolus
+        case exceedsMaxBolus
+        
+        var errorDescription: String? {
+            switch self {
+            case .invalidBolus:
+                return NSLocalizedString("Invalid Bolus Amount", comment: "Bolus error description: invalid bolus amount.")
+            case .missingMaxBolus:
+                return NSLocalizedString("Missing maximum allowed bolus in settings", comment: "Bolus error description: missing maximum bolus in settings.")
+            case .exceedsMaxBolus:
+                return NSLocalizedString("Exceeds maximum allowed bolus in settings", comment: "Bolus error description: bolus exceeds maximum bolus in settings.")
+            }
+        }
+    }
     
-    func handleCarbAction(_ action: CarbAction) async throws {
-        let candidateCarbEntry = try action.toValidCarbEntry(defaultAbsorptionTime: carbStore.defaultAbsorptionTimes.medium,
-                                                                  minAbsorptionTime: LoopConstants.minCarbAbsorptionTime,
-                                                                  maxAbsorptionTime: LoopConstants.maxCarbAbsorptionTime,
-                                                                  maxCarbEntryQuantity: LoopConstants.maxCarbEntryQuantity.doubleValue(for: .gram()),
-                                                                  maxCarbEntryPastTime: LoopConstants.maxCarbEntryPastTime,
-                                                                  maxCarbEntryFutureTime: LoopConstants.maxCarbEntryFutureTime
-        )
+    //Carb Entry
+    
+    func deliverCarbs(amountInGrams: Double, absorptionTime: TimeInterval?, foodType: String?, startDate: Date?) async throws {
+        
+        let absorptionTime = absorptionTime ?? carbStore.defaultAbsorptionTimes.medium
+        if absorptionTime < LoopConstants.minCarbAbsorptionTime || absorptionTime > LoopConstants.maxCarbAbsorptionTime {
```

Comment:
I think validation and error reporting for remote commands should remain in ServiceManager. Local commands tend to be validated directly in the local UIs before being commanded.